packages/plugin-backgroundcolor/src/BackgroundColor.ts

```diff
@@ -0,0 +1,164 @@
+};
+
+export class BackgroundColor<T extends BackgroundColorConfig = BackgroundColorConfig> extends JWPlugin<T> {
+    static hasColor(node: VNode): boolean;
+    static hasColor(backgroundColor: string, node: VNode): boolean;
+    static hasColor(backgroundColor: string | VNode, node?: VNode): boolean {
+        if (backgroundColor instanceof VNode) {
+            node = backgroundColor;
+        }
+        const nodeBackgroundColor = getStyle(node, 'background-color');
+        if (!nodeBackgroundColor || backgroundColor instanceof VNode) {
+            return !!nodeBackgroundColor;
+        } else {
+            return nodeBackgroundColor === backgroundColor;
+        }
+    }
+    static dependencies = [Inline];
+    configuration = { ...defaultConfiguration, ...this.configuration };
+    commands = {
+        colorBackground: {
+            handler: this.colorBackground,
+        },
+        uncolorBackground: {
+            handler: this.uncolorBackground,
+        },
+    };
+    readonly loadables: Loadables<Keymap> = {
+        shortcuts: [
+            {
+                pattern: 'CTRL+H',
+                commandId: 'colorBackground',
+                // TODO: use dialog to get params
+                commandArgs: {
+                    color: 'yellow',
+                },
+            },
+            {
+                pattern: 'CTRL+SHIFT+H',
+                commandId: 'uncolorBackground',
+            },
+        ],
+    };
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    /**
+     * Apply the given background color to the range.
+     *
+     * @param params
+     */
+    colorBackground(params: ColorParams): void {
+        const color = params.color;
+
+        if (params.context.range.isCollapsed()) {
+            // Set the style cache.
+            const inline = this.dependencies.get(Inline);
+            const currentCache = inline.styleCache || {};
+            inline.styleCache = { ...currentCache, 'background-color': color };
+        } else {
+            for (const node of this._highestSelectedNodes(params.context.range)) {
+                // Apply the style to the node or its first format.
+                setStyle(this._nodeOrFirstFormat(node), 'background-color', color);
+
+                // If there are ancestors of this node whose children all have
+                // this style, style these ancestors instead of their
+                // descendants.
+                let parent = node.parent;
+                while (parent?.children().every(child => BackgroundColor.hasColor(color, child))) {
+                    setStyle(parent, 'background-color', color);
+                    for (const child of parent.children()) {
+                        removeStyle(child, 'background-color');
+                    }
+                    parent = parent.parent;
+                }
```

Comment:
Actually, the gspec acts weirdly. It sets the color on the container if you set the color its last character. It seems to be the only rule. You don't even need to select everything. Coloring the last character of a container colors the container. Not saying it's great though, but I have the feeling that going futher than the direct parent is overengineering here.

Reply:
I don't know. I think this is better than the gspec. It's easy to target the highest and it's safer. Don't forget we don't know in advance what we're going to parse...